File: web/src/adapters/remoteFtp.ts
import { NormalizedRow, NumericField, toFloat } from "./normalized";

export const KEY = "remote-ftp" as const;

/** Maps the published `sensor_raw` / `sensor_9am` columns to the normalized model. */
export const HEADER_MAP: Record<string, keyof NormalizedRow> = {
  "date": "ts",
  "addr": "deviceId",
  "number": "sensorNumber",
  "battery1": "battery_v",
  "battery1[V]": "battery_v",
  "bulk_ec": "ec_bulk_dsm",
  "bulk_ec[dS/m]": "ec_bulk_dsm",
  "vwc": "vwc_pct",
  "vwc[%]": "vwc_pct",
  "soil_temp": "temperature_c",
  "soil_temp[℃]": "temperature_c",
  "air_temp": "air_temp_c",
  "air_tmp": "air_temp_c",
  "air_tmp[℃]": "air_temp_c",
  "air_temp[℃]": "air_temp_c",
  "precip_1h": "precip_1h_mm",
  "precip_1h[mm]": "precip_1h_mm",
  "sunshine_1h": "sunshine_1h_h",
  "sunshine_1h[h]": "sunshine_1h_h",
};

const STRING_FIELDS = new Set<keyof NormalizedRow>(["ts", "deviceId", "sensorNumber"]);

export function toNormalized(values: string[][]): NormalizedRow[] {
  if (!values.length) return [];
  const header = values[0];
  const idx: { col: number; field: keyof NormalizedRow }[] = [];
  header.forEach((name, col) => {
    const field = HEADER_MAP[name];
    if (field) idx.push({ col, field });
  });
  const out: NormalizedRow[] = [];
  for (let r = 1; r < values.length; r++) {
    const row = values[r];
    if (!row || row.length === 0) continue;
    const obj: NormalizedRow = {};
    for (const { col, field } of idx) {
      const cell = col < row.length ? row[col] : "";
      if (STRING_FIELDS.has(field)) {
        const s = String(cell ?? "").trim();
        if (s) (obj as Record<string, unknown>)[field] = s;
      } else {
        const n = toFloat(cell);
        if (n !== undefined) (obj as Record<string, unknown>)[field as NumericField] = n;
      }
    }
    if (obj.ts === undefined && obj.deviceId === undefined) continue;
    out.push(obj);
  }
  return out;
}

File: scripts/adapters/remote_ftp.py
"""Remote FTP data-source adapter (SPEC §4.2, remote-ftp).

Reads the published `sensor_raw` / `sensor_9am` schema produced by
`scripts.sensor_parser.to_published`:
    date, siteId, addr, number, battery1, battery2, bulk_ec, vwc, soil_temp
"""

from __future__ import annotations

from .normalized import NormalizedRow, to_float

KEY = "remote-ftp"

HEADER_MAP = {
    "date": "ts",
    "addr": "device_id",
    "battery1": "battery_v",
    "battery1[V]": "battery_v",
    "bulk_ec": "ec_bulk_dsm",
    "bulk_ec[dS/m]": "ec_bulk_dsm",
    "vwc": "vwc_pct",
    "vwc[%]": "vwc_pct",
    "soil_temp": "temperature_c",
}

STRING_FIELDS = {"ts", "device_id"}


def to_normalized(values: list[list[str]]) -> list[NormalizedRow]:
    if not values:
        return []
    header = values[0]
    idx = {name: i for i, name in enumerate(header) if name in HEADER_MAP}
    out: list[NormalizedRow] = []
    for row in values[1:]:
        if not row:
            continue
        row_obj = NormalizedRow()
        for raw_name, col in idx.items():
            field = HEADER_MAP[raw_name]
            cell = row[col] if col < len(row) else ""
            if field in STRING_FIELDS:
                setattr(row_obj, field, (str(cell).strip() or None))
            else:
                setattr(row_obj, field, to_float(cell))
        if row_obj.ts is None and row_obj.device_id is None:
            continue
        out.append(row_obj)
    return out
